src/shims/foreign_items.rs

```diff
@@ -185,25 +220,31 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         }
 
         // Third: functions that return.
-        if this.emulate_foreign_item_by_name(link_name, abi, args, dest, ret)? {
-            trace!("{:?}", this.dump_place(**dest));
-            this.go_to_block(ret);
+        if let Some(returns) = this.emulate_foreign_item_by_name(link_name, abi, args, dest, ret)? {
+            if returns {
+                trace!("{:?}", this.dump_place(**dest));
+                this.go_to_block(ret);
+            }
+        } else if let Some(body) = this.lookup_non_external_foreign_item(abi, link_name_sym)? {
+            return Ok(Some(body));
+        } else {
+            throw_unsup_format!("can't call foreign function: {}", link_name);
         }
 
         Ok(None)
     }
 
-    /// Emulates calling a foreign item using its name, failing if the item is not supported.
-    /// Returns `true` if the caller is expected to jump to the return block, and `false` if
-    /// jumping has already been taken care of.
+    /// Emulates calling a foreign item using its name, returns `None` if the item is not supported.
+    /// Returns `Some(true)` if the caller is expected to jump to the return block, and `Some(false)`
+    /// if jumping has already been taken care of.
```

Comment:
I feel like we should introduce a custom `enum` here instead of assigning increasing complicated semantics to more-or-less arbitrary combinaitons of standard library types.